src/Features/CSharp/Portable/SignatureHelp/TupleConstructionSignatureHelpProvider.cs

```diff
@@ -0,0 +1,192 @@
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.SignatureHelp;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Roslyn.Utilities;
+using System.Collections.Immutable;
+
+namespace Microsoft.CodeAnalysis.CSharp.SignatureHelp
+{
+    [ExportSignatureHelpProvider("TupleSignatureHelpProvider", LanguageNames.CSharp), Shared]
+    internal class TupleConstructionSignatureHelpProvider : AbstractCSharpSignatureHelpProvider
+    {
+        private static readonly Func<TupleExpressionSyntax, SyntaxToken> s_getOpenToken = e => e.OpenParenToken;
+        private static readonly Func<TupleExpressionSyntax, SyntaxToken> s_getCloseToken = e => e.CloseParenToken;
+        private static readonly Func<TupleExpressionSyntax, IEnumerable<SyntaxNodeOrToken>> s_getArgumentsWithSeparators = e => e.Arguments.GetWithSeparators();
+        private static readonly Func<TupleExpressionSyntax, IEnumerable<string>> s_getArgumentNames = e => e.Arguments.Select(a => a.NameColon?.Name.Identifier.ValueText ?? string.Empty);
+
+        public override SignatureHelpState GetCurrentArgumentState(SyntaxNode root, int position, ISyntaxFactsService syntaxFacts, TextSpan currentSpan, CancellationToken cancellationToken)
+        {
+            TupleExpressionSyntax expression;
+            if (TryGetTupleExpression(SignatureHelpTriggerReason.InvokeSignatureHelpCommand,
+                root, position, syntaxFacts, cancellationToken, out expression) &&
+                 currentSpan.Start == expression.SpanStart)
+            {
+                return CommonSignatureHelpUtilities.GetSignatureHelpState(expression, position,
+                    getOpenToken: s_getOpenToken,
+                    getCloseToken: s_getCloseToken,
+                    getArgumentsWithSeparators: s_getArgumentsWithSeparators,
+                    getArgumentNames: s_getArgumentNames);
+            }
+
+            ParenthesizedExpressionSyntax parenthesizedExpression;
+            if (TryGetParenthesizedExpression(SignatureHelpTriggerReason.InvokeSignatureHelpCommand, 
+                root, position, syntaxFacts, cancellationToken, out parenthesizedExpression))
+            {
+                // This could only have parsed as a parenthesized expression in these two cases:
+                // ($$)
+                // (name$$)
+                string name = 0.ToString(); // This causes the controller to match against the 0th tuple member
+                if (parenthesizedExpression.Expression is IdentifierNameSyntax)
+                {
+                    name = ((IdentifierNameSyntax)parenthesizedExpression.Expression).Identifier.ValueText;
+                }
+
+                return new SignatureHelpState(
+                    argumentIndex: 0,
+                    argumentCount: 0,
+                    argumentName: name,
```

Comment:
tuples don't have reorderable named arguments.  i.e. if i have:

```
(int a, int b) = (b: 0, a: 0);
```
Then even in teh "b:0" case, we're still in the 0th argument that corresponds to 'int a'.